src/repr/src/row.rs

```diff
@@ -1193,6 +1350,172 @@ impl RowPacker<'_> {
         })
     }
 
+    /// Pushes a `DatumRange` that is considered empty.
+    ///
+    /// - To construct other ranges, use [`Self::push_range`],
+    ///   [`Self::push_range_with`].
+    pub fn push_empty_range(&mut self) {
+        self.row.data.push(Tag::Range.into());
+        // Untagged bytes only contains the `RANGE_EMPTY` flag value.
+        self.row.data.extend_from_slice(&1u64.to_le_bytes());
+        self.row.data.push(range::RANGE_EMPTY);
+    }
+
+    /// Pushes a `DatumRange` built from the specified arguments.
+    ///
+    /// - If an error occurs, the packer's state will be unchanged.
+    /// - Infinite boundaries can be expressed using [`Datum::Null`]. Note that
+    ///   this differs from [`Self::push_range_with`], which requires ifinite
+    ///   boundaries be expressed with [`RangeBoundDesc`].
+    /// - To construct an empty range, use [`Self::push_empty_range`]. Note
+    ///   that it is possible to generate semantically empty ranges with this
+    ///   function, but they will _not_ be output as empty ranges.
+    /// - To construct ranges without providing the `Datum`s directly, see
+    ///   [`Self::push_range_with`].
+    ///
+    /// # Panics
+    /// - If `lower` and `upper` are both non-null and are of different types.
+    pub fn push_range<'a, D>(
+        &mut self,
+        lower: D,
+        lower_inc: bool,
+        upper: D,
+        upper_inc: bool,
+    ) -> Result<(), RangeError>
+    where
+        D: Borrow<Datum<'a>>,
+    {
+        let lower = lower.borrow();
+        let upper = upper.borrow();
+
+        self.push_range_with(
+            RangeBoundDesc {
+                inf: matches!(lower, Datum::Null),
+                inc: lower_inc,
+            },
+            RangeBoundDesc {
+                inf: matches!(upper, Datum::Null),
+                inc: upper_inc,
+            },
+            |row| {
+                for bound in [lower, upper] {
+                    if !matches!(bound, Datum::Null) {
```

Comment:
Consumers of this can distinguish between lower and upper by checking the `RangeBoundDesc`s?

Reply:
Not totally sure I parse your question. When consuming range datums that are pushed with this method, one needs to check the `u8` flag at the beginning of the untagged bytes to determine which datums are alongside it. The convention is if the lower is present present, it's placed before the upper. The details of this are a little more explicit in `push_range_with`.